### PR TITLE
Add CommentPruner to Prow to save API tokens when deleting stale comments.

### PR DESCRIPTION
**Pull request by cjwagner**
This is a prow parallel to mungegithub's comment-deleter munger.

/cc @rmmh

### CODE REVIEW

File: prow/commentpruner/commentpruner.go [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+	repo   string
+	number int
+
+	ghc githubClient
+	log *logrus.Entry
+
+	comments []github.IssueComment
+
+	once sync.Once  // Used to fetch the IssueComments once.
+	lock sync.Mutex // Protects `comments`.
```
Comment (kargakis) [inline prow/commentpruner/commentpruner.go]: As a convention locks are added on top of the fields they protect.

**Comment by spxtr** [comment]
What plugins need this?

**Comment by cjwagner** [comment]
> What plugins need this?

This is for some of the plugins that are being migrated from mungegithub. It will also be used by the release-note plugin.

**Comment by spxtr** [comment]
Can you go into more detail about why plugins need this? I'm worried that it will actually cause *more* token usage, because instead of plugins smartly deciding when they need to delete comments, they will instead call `PruneComments` on every webhook or something.

That said, if it makes migrating mungegithub -> prow easier then I'm all for it. It's a well-designed package :)

**Comment by cjwagner** [comment]
Plugins should still smartly decide to use `PruneComments` and should not use it on every webhook event. That being said, once all the mungers are migrated to Prow, some webhook events that hook receives will be likely to have multiple handlers request comment pruning. Without this client, the handlers would each have to fetch the comments which would use more API tokens.
Even an HTTP cache shared by the handlers wouldn't completely minimize token usage like this client does because each deletion causes the list of issues on github to change, invalidating the cache entry.

File: prow/commentpruner/commentpruner.go (hunk 1 above)
Comment (cjwagner) [reply prow/commentpruner/commentpruner.go]: That makes sense. Fixed.

File: prow/commentpruner/commentpruner.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+		org:    org,
+		repo:   repo,
+		number: number,
+
+		ghc: ghc,
+		log: log,
+	}
+}
+
+// PruneComments fetchest issue comments if they have not yet been fetched for this webhook event
```
Comment (kargakis) [inline prow/commentpruner/commentpruner.go]: typo

File: prow/commentpruner/commentpruner.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+import (
+	"sync"
+
+	"github.com/sirupsen/logrus"
+
+	"k8s.io/test-infra/prow/github"
+)
+
+// bufSize is the size of the sliding window and subsequently the cache size.
+const bufSize = 100
```
Comment (rmmh) [inline prow/commentpruner/commentpruner.go]: unused

File: prow/commentpruner/commentpruner.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+		log: log,
+	}
+}
+
+// PruneComments fetchest issue comments if they have not yet been fetched for this webhook event
+// and then deletes any indicated by the func 'shouldPrune'.
+func (c *EventClient) PruneComments(shouldPrune func(github.IssueComment) bool) {
+	c.once.Do(func() {
+		c.lock.Lock()
+		defer c.lock.Unlock()
```
Comment (rmmh) [inline prow/commentpruner/commentpruner.go]: the locking here is unnecessary-- once provides its own lock

File: prow/hook/events.go [hunk 5]
```diff
@@ -39,6 +40,13 @@ func (s *Server) handleReviewEvent(l *logrus.Entry, re github.ReviewEvent) {
 			pc.Logger = l.WithField("plugin", p)
 			pc.Config = s.ConfigAgent.Config()
 			pc.PluginConfig = s.Plugins.Config()
+			pc.CommentPruner = commentpruner.NewEventClient(
```
Comment (rmmh) [inline prow/hook/events.go]: this is kind of repetitive but I guess the whole file is

File: prow/commentpruner/commentpruner.go (hunk 4 above)
Comment (kargakis) [reply prow/commentpruner/commentpruner.go]: The lock is there likely because `once` does not protect a second thread from accessing the comment slice.

Comment (cjwagner) [reply prow/commentpruner/commentpruner.go]: I think @rmmh  is right. All threads block on `c.once.Do` until the function has returned. See [here](https://golang.org/pkg/sync/#Once.Do).

File: prow/commentpruner/commentpruner.go (hunk 2 above)
Comment (cjwagner) [reply prow/commentpruner/commentpruner.go]: Fixed.

File: prow/commentpruner/commentpruner.go (hunk 3 above)
Comment (cjwagner) [reply prow/commentpruner/commentpruner.go]: Fixed.

**Comment by cjwagner** [comment]
Comments addressed. PTAL

**Comment by cjwagner** [comment]
Addressed comments and updated to only consider bot comments for pruning so that handlers don't have to include that check themselves (plugins should never need to delete human made comments).

**Pull request merged**